larana/OpticalDetector/OpHitFinder/RiseTimeTools/RiseTimeGaussFit_tool.cc

```diff
@@ -0,0 +1,136 @@
+
+namespace pmtana {
+
+  class RiseTimeGaussFit : RiseTimeCalculatorBase {
+
+  public:
+    //Configuration parameters
+    struct Config {
+
+      fhicl::Atom<double> MinAmp{fhicl::Name("MinAmp")};
+      fhicl::Atom<double> InitSigma{fhicl::Name("InitSigma")};
+      fhicl::Atom<int> Nbins{fhicl::Name("Nbins")};
+      fhicl::Atom<double> Tolerance{fhicl::Name("Tolerance")};
+    };
+
+    // Default constructor
+    explicit RiseTimeGaussFit(art::ToolConfigTable<Config> const& config);
+
+    // Method to calculate the OpFlash t0
+    double RiseTime(const pmtana::Waveform_t& wf_pulse,
+                    const pmtana::PedestalMean_t& ped_pulse,
+                    bool _positive) const override;
+    // Method to fit the first local max of the wvf above fixed threshold
+    int findFirstMax(const std::vector<double>& arr, int low, int high, double threshold) const;
+
+  private:
+    double fMinAmp;
+    double fInitSigma;
+    double fTolerance;
+    int fNbins;
+  };
+
+  RiseTimeGaussFit::RiseTimeGaussFit(art::ToolConfigTable<Config> const& config)
+    : fMinAmp{config().MinAmp()}
+    , fInitSigma{config().InitSigma()}
+    , fTolerance{config().Tolerance()}
+    , fNbins{config().Nbins()}
+  {}
+
+  double RiseTimeGaussFit::RiseTime(const pmtana::Waveform_t& wf_pulse,
+                                    const pmtana::PedestalMean_t& ped_pulse,
+                                    bool _positive) const
+  {
+
+    // Pedestal-subtracted pulse
+    std::vector<double> wf_aux(ped_pulse);
+    if (_positive) {
+      for (size_t ix = 0; ix < wf_aux.size(); ix++) {
+        wf_aux[ix] = ((double)wf_pulse[ix]) - wf_aux[ix];
+      }
+    }
+    else {
+      for (size_t ix = 0; ix < wf_aux.size(); ix++) {
+        wf_aux[ix] = wf_aux[ix] - ((double)wf_pulse[ix]);
+      }
+    }
+
+    // Find first local maximum
+    size_t first_max = findFirstMax(wf_aux, 0, wf_aux.size(), fMinAmp);
+
+    // Create & fill th1
+    TH1F* h_aux = new TH1F("aux", "aux", wf_aux.size(), -0.5, wf_aux.size() - 0.5);
+    for (long unsigned int j = 0; j < wf_aux.size(); j++)
+      h_aux->SetBinContent(j + 1, wf_aux[j]);
+
+    // Function & initial values for the fit, ROOT Gaussian:  [p0]*e**(-0.5 (x-[p1])**2 / [p2]**2)
+    TF1* f = new TF1("f", "gaus", double(first_max) - fNbins, double(first_max) + fNbins);
+    std::vector<double> p_init = {wf_aux[first_max], double(first_max), fInitSigma};
+    for (long unsigned int p = 0; p < wf_aux.size(); p++)
+      f->SetParameter(p, p_init[p]);
+
+    // Fit
+    h_aux->Fit(f, "q", "SAME", first_max - fNbins, first_max + fNbins);
+    double t_fit = f->GetParameter(1);
+    double peak_time;
+    if (
+      std::abs(t_fit - first_max) <
+      fTolerance) { //check fit is close in distance to the original max, use max peak as time otherwise
+      peak_time = t_fit;
+    }
+    else {
+      peak_time = first_max;
+      std::cout << "No good fit found, keeping 1st max bin instead" << std::endl;
```

Comment:
Is it possible to replace the use of `std::cout` with the message-facility?  By using `std::cout`, there is no way to suppress the printout to STDOUT.  If, instead, you use something like `mf::LogInfo("Fit") << "No good fit..."`, then users have the option of suppressing the message if it's not desired.